invenio_app_ils/ui/src/invenio_app_ils/common/api/loans/loan.js

```diff
@@ -0,0 +1,119 @@
+import { http } from '../base';
+import { toISO } from '../date';
+import { DateTime } from 'luxon';
+
+const loanURL = '/circulation/loans/';
+const get = loanPid => http.get(`${loanURL}${loanPid}`);
+const getLoanReplaceItemUrl = loanPid => `${loanURL}/${loanPid}/replace-item`;
+
+const postAction = (
+  url,
+  pid,
+  loan,
+  transactionUserPid,
+  transactionLocationPid
+) => {
+  const now = DateTime.local();
+  const payload = {
+    transaction_user_pid: transactionUserPid,
+    patron_pid: loan.patron_pid,
+    transaction_location_pid: transactionLocationPid,
+    transaction_date: toISO(now),
+  };
+
+  if ('item_pid' in loan) {
+    payload['item_pid'] = loan.item_pid;
+  } else if ('document_pid' in loan) {
+    payload['document_pid'] = loan.document_pid;
+  } else {
+    throw new Error(
+      `No 'item_pid' or 'document_pid' attached to loan '${pid}'`
+    );
+  }
+  return http.post(url, payload);
+};
+
+const assignItemToLoan = (itemPid, loanPid) => {
+  const url = getLoanReplaceItemUrl(loanPid);
+  const payload = { item_pid: itemPid };
+  return http.post(url, payload);
+};
+
+class Loans {
+  constructor(build) {
+    this.documentQuery = build.documentQuery ? build.documentQuery : '';
+    this.itemQuery = build.itemQuery ? build.itemQuery : '';
+    this.patronQuery = build.patronQuery ? build.patronQuery : '';
+    this.stateQuery = build.stateQuery ? build.stateQuery : '';
+  }
+
+  qs() {
+    return `${this.documentQuery}${this.itemQuery}${this.patronQuery}${
+      this.stateQuery
+    }`.replace(' AND ', '');
+  }
+
+  static get Builder() {
+    class Builder {
+      withDocPid(documentPid) {
+        this.documentQuery = documentPid
+          ? ` AND document_pid:${Builder.paramToQuery(documentPid)}`
+          : '';
+        return this;
+      }
+
+      withItemPid(itemPid) {
+        this.itemQuery = itemPid
+          ? ` AND item_pid:${Builder.paramToQuery(itemPid)}`
+          : '';
+        return this;
+      }
+
+      withPatronPid(patronPid) {
+        this.patronQuery = patronPid
+          ? ` AND patron_pid:${Builder.paramToQuery(patronPid)}`
+          : '';
+        return this;
+      }
+
+      withState(state) {
```

Comment:
Just a suggestion here, all these functions look the same, I would try to use an options pattern. We can possibly replace `withDocPid`, `withItemPid`, `withPatronPid`, `withState` with something like:

```
params = {
    document_pid: documentPid,
    item_pid: itemPid
}

withParam(params) {
    qry = '';
    for (let val in params) {
        qry = `${qry} AND ${val}:${Builder.paramToQuery(params[val])}`
    }
}
```
Like this whenever we want to extend further our queries we just pass the param with the value we want to search for, we won't have to implement yet another identical function.

Reply:
I also prefer the Builder, very readable and "usable"